### DISCUSSION TITLE
execute_action fails

### DISCUSSION THREAD

**Discussion by bakerwy** [fumitoh/modelx#112] [Q&A]
Hi,

I built model and finally tried to run the model with the method below link
since it uses too much memory

[https://lifelib.io/libraries/savings/savings_example3.html](https://lifelib.io/libraries/savings/savings_example3.html)

the contents of the link has three steps.

1. generate model
2. save the model
3. execute the model with ipyparallel

When I execute the model with the code in step3, the error occured,

```python
code = """
import modelx as mx
m = mx.read_model('C:\modelx\models\generated3')
m.Projection.model_point_table = model_point
m.execute_actions(m.Projection.actions)
result = m.Projection.BEL_result()
"""

rc[:].execute(code, block=True)
```

FYI, generated3 is generated model from step1.

The error message said..

```
AttributeError:

[1:execute]:
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
~\AppData\Local\Temp\ipykernel_17732\2127489956.py in <cell line: 4>()
      2 m = mx.read_model('C:\modelx\models\generated3')
      3 m.Projection.model_point_table = model_point
----> 4 m.execute_actions(m.Projection.actions)
      5 result = m.Projection.BEL_result()

C:\modelx\WPy64-31020\python-3.10.2.amd64\lib\site-packages\modelx\core\space.py in __getattr__(self, name)
    306             return self._impl.get_attr(name)
    307         else:
--> 308             raise AttributeError  # Must return AttributeError for hasattr
    309
    310     def __dir__(self):

.....
```

it seems it has problem with `m.Projection.actions` which was generated.
I tested in step 1. generated model with the code below and it works well.

```python
actions = model.generate_actions([model.Projection.BEL_result.node()])
model.execute_actions(actions)
model.Projection.BEL_result()
```

I think it has some problems when I saved the model.
I tried several times and it has same errors in step3.

Is there things to do when saving models?
and would you let me know where can I find information about generated model in the saved model? 
I can't find information about generated model in `__init__` file in Projection folder.

**Comment by fumitoh** [comment]
Further tests may be needed to identify the cause of the error.

```python
actions = model.generate_actions([model.Projection.BEL_result.node()])
model.clear_all()
model.execute_actions(actions)
model.Projection.BEL_result()
```

Does the above code run successfully? The key difference from your test is the inclusion of `model.clear_all()`. 

If this is successful, then let's proceed with the next test in a new Python session:

First, run:
```python
actions = model.Projection.actions =  model.generate_actions([model.Projection.BEL_result.node()])
model.write("test_actions")
```

Then, start another Python session and run the following:
```python
model = mx.read_model("test_actions")
model.execute_actions(model.Projection.actions)
model.Projection.BEL_result()
```

Does this run successfully? If it fails, compare the contents of `actions` from the first session with `model.Projection.actions` from the second session. Do you notice any differences?

`model.Projection.actions` is a nested list. It is stored as binary data in the `data.pickle` file within the `_data` folder, so it cannot be checked directly.

**Comment by bakerwy** [comment]
I tried your code and successfully saved file and checked the code below and works well.

model = mx.read_model("C:/modelx/models/test_action")
model.execute_actions(model.Projection.actions)
model.Projection.BEL_result()

**but new problem faced.**

I executed action with 'test_action' file with new model point but model point is not updated. The tested codes are below

m = mx.read_model('C:/modelx/models/test_action')
m.Projection.model_point_table = new_model_point
m.execute_actions(m.Projection.actions)
result = m.Projection.BEL_result()

When I tried the code "m.execute_actions(m.Projection.actions)" any calculations are not executed. also "result" is the result that I previously saved result when the model generated action.

Could you give me further helps?

**Reply by fumitoh** [reply]
You can check:
* What is assigned to `Projection.actions`, in the models before and after the write/read.
* What happens withought `execute_actions`, but with   the `m.Projection.model_point_table = new_model_point` assignment. If `result` is not updated for the new model points, then it's not a problem of `execute_actions`.

**Reply by bakerwy** [reply]
Okay I will try again thanks :)
